### PR TITLE
CRM: Let Woo invoices use crm reference number

### PR DESCRIPTION
**Pull request by gogdzl**
Fixes https://github.com/Automattic/zero-bs-crm/issues/3457

## Proposed changes:

* This PR adds a setting to enable invoices generated by WooCommerce to use the Jetpack CRM invoice reference system.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
https://github.com/Automattic/zero-bs-crm/issues/3457

## Testing instructions:
**Initial Setup and Verification of WooSync:**

1. **Verify WooSync Functionality:**
   - Navigate to the WooSync hub.
   - Confirm that the status displays `Status: Sync Completed`.

2. **Enable Invoice Generation:**
   - If invoice generation is not active, switch to the `trunk` branch and activate the feature at `/wp-admin/admin.php?page=zerobscrm-plugin-settings&tab=woosync`.

3. **Create a WooCommerce Order:**
   - Place a new order in WooCommerce.

4. **Verify Transaction and Invoice Creation:**
   - Ensure both the transaction and invoice are correctly generated.

**Testing the New PR Setting:**

1. **Switch to the PR Branch:**
   - Change to the branch `update/crm/3457-let-woo-invoices-use-crm-reference-number`.

2. **Configure WooCommerce Settings:**
   - Access Woo settings at `/wp-admin/admin.php?page=zerobscrm-plugin-settings&tab=woosync`.
   - Enable `Use the Jetpack CRM invoice reference system for invoices generated by WooCommerce` and `Create Invoices from WooCommerce Orders`.

3. **Set Up Invoicing Reference System:**
   - Go to the Invoicing settings page at `/wp-admin/admin.php?page=zerobscrm-plugin-settings&tab=invbuilder`.
   - Configure `Auto-generated reference` by adding a prefix, setting the next number, and specifying a suffix.

4. **Add Another WooCommerce Order:**
   - Place a new order and confirm that the invoice generated reflects the configured auto-generated reference system.

5. **Create a Manual Invoice:**
   - Navigate to `wp-admin/admin.php?page=zbs-add-edit&action=edit&zbstype=invoice` and create a new invoice.
   - Verify that the reference number for the new invoice follows the auto-generated system, incrementing sequentially from the last WooCommerce order.

6. **Final Verification:**
   - Place one more WooCommerce order to ensure the invoice reference system is consistently applied across all new invoices.

## Documentantion changes

We will need to update our CRM Knowledge Base to reflect this change. It is necessary to update our WooSync page to inform users that we now offer a new setting and explain how it works. Additionally, it would be beneficial to mention that this feature can aid in compliance with EU regulations regarding sequential invoices. This is achievable if the user utilizes the new setting and exclusively relies on the CRM for invoice creation. Ensure to caution the user that this functionality is only effective when Auto generation is enabled in our Invoicing Settings page.

Pinging @StefMattana

### CODE REVIEW

File: projects/plugins/crm/includes/ZeroBSCRM.DAL3.Obj.Invoices.php [hunk 1]
```diff
@@ -1794,7 +1799,20 @@ public function addUpdateInvoice($args=array()){
...
-            
+					// If we are using our CRM reference id (table field id_override) system, we should generate a new invoice number.
+					if ( isset( $data['woo_use_crm_id'] ) && $data['woo_use_crm_id'] === true ) {
+						$ref_type = $zbs->settings->get( 'reftype' );
+						// We can only generate it if autonumber is set
+						if ( $ref_type === 'autonumber' ) {
+							$next_number                 = $zbs->settings->get( 'refnextnum' );
+							$prefix                      = $zbs->settings->get( 'refprefix' );
+							$suffix                      = $zbs->settings->get( 'refsuffix' );
+							$dataArr['zbsi_id_override'] = $prefix . $next_number . $suffix; // phpcs:ignore WordPress.NamingConventions.ValidVariableName.VariableNotSnakeCase
```
Comment (coder-karen) [inline projects/plugins/crm/includes/ZeroBSCRM.DAL3.Obj.Invoices.php]:
```suggestion
							$next_number                 = $zbs->settings->get( 'refnextnum' );
							$dataArr['zbsi_id_override'] = $zbs->settings->get( 'refprefix' ) . $next_number . $zbs->settings->get( 'refsuffix' ); // phpcs:ignore WordPress.NamingConventions.ValidVariableName.VariableNotSnakeCase
```

Could we remove two of those variables as we don't use them anywhere else?

File: projects/plugins/crm/modules/woo-sync/admin/settings/main.page.php [hunk 2]
```diff
@@ -56,10 +56,11 @@ function jpcrm_settings_page_html_woosync_main() {
...
+		$updatedSettings['wccrminvreference'] = ! empty( $_POST['wpzbscrm_wccrminvreference'] ); // phpcs:ignore WordPress.NamingConventions.ValidVariableName.VariableNotSnakeCase, WordPress.Security.NonceVerification.Missing
+		$updatedSettings['wcprod']            = ! empty( $_POST['wpzbscrm_wcprod'] ); // phpcs:ignore WordPress.NamingConventions.ValidVariableName.VariableNotSnakeCase, WordPress.Security.NonceVerification.Missing
+		$updatedSettings['wcport']            = ! empty( $_POST['wpzbscrm_wcport'] ) ? preg_replace( '/\s*,\s*/', ',', sanitize_text_field( $_POST['wpzbscrm_wcport'] ) ) : ''; // phpcs:ignore WordPress.NamingConventions.ValidVariableName.VariableNotSnakeCase, WordPress.Security.NonceVerification.Missing, WordPress.Security.ValidatedSanitizedInput.MissingUnslash
+		$updatedSettings['wcacc']             = ! empty( $_POST['wpzbscrm_wcacc'] ); // phpcs:ignore WordPress.NamingConventions.ValidVariableName.VariableNotSnakeCase, WordPress.Security.NonceVerification.Missing
```
Comment (coder-karen) [inline projects/plugins/crm/modules/woo-sync/admin/settings/main.page.php]:
```suggestion
		// phpcs:disable WordPress.NamingConventions.ValidVariableName.VariableNotSnakeCase, WordPress.Security.NonceVerification.Missing
		$updatedSettings['wcinv']             = ! empty( $_POST['wpzbscrm_wcinv'] );
		$updatedSettings['wccrminvreference'] = ! empty( $_POST['wpzbscrm_wccrminvreference'] );
		$updatedSettings['wcprod']            = ! empty( $_POST['wpzbscrm_wcprod'] );
		$updatedSettings['wcport']            = ! empty( $_POST['wpzbscrm_wcport'] ) ? preg_replace( '/\s*,\s*/', ',', sanitize_text_field( wp_unslash( $_POST['wpzbscrm_wcport'] ) ) ) : ''; // phpcs:ignore WordPress.Security.ValidatedSanitizedInput.MissingUnslash - wcport values can be saved with slashes.
		$updatedSettings['wcacc']             = ! empty( $_POST['wpzbscrm_wcacc'] );
		 // phpcs:enable WordPress.NamingConventions.ValidVariableName.VariableNotSnakeCase, WordPress.Security.NonceVerification.Missing
```
Perhaps this is slightly cleaner? With the MissingUnslash sniff, I believe `wcport` is related to the list of Jetpack CRM contact fields that can be edited via WooCommerce My Account. I added one with a slash and it does save so it does look like the sniff needs to stay - maybe with an additional note.

**Review by coder-karen — DISMISSED** [review]
Works well for me!

I added some non blocker suggestions too.

**Review by coder-karen — APPROVED** [review]
LGTM!

**Comment by StefMattana** [comment]
Doc request: https://github.com/Automattic/jpop-quill/issues/178

**Pull request merged**